### ISSUE TITLE
Windows ARM64 Support

### ISSUE THREAD

**Issue by Glavo** [oracle/graal#9215]
Since the release of Qualcomm Snapdragon X Elite, Windows ARM64 has received more attention. I wonder if there are plans for GraalVM to support Windows ARM64 platform now?

**Comment by selhagani** [comment]
Hi @Glavo, thank you for reaching out to us!

Support for Windows ARM64 is not planned for GraalVM currently.

**Comment by hohwille** [comment]
So you are saying that you neither plan cross compilation (see #8350 or #407) nor do you want to provide an ARM release of GraalVM.
In other words ARM users are locked out of GraalVM.
What a pity.

**Comment by hohwille** [comment]
Please be aware the ARM is the new big thing in the laptop market as the Snapdragon X CPU family is so much more energy efficient and like MacOS where the revolution has already taken place and x86 is dead, on Windows and Linux it is just about happening. I bought this ARM laptop because I am an OSS developer and want to build and test my releases of software products I am maintaining also for the ARM target architecture. Currently windows on ARM feels a little bleeding edge and some software does not even run with emulation so it is not yet ideal for the mainstream user but this is only because software vendors like you and me do not properly support it yet and need to do their homework.

**Comment by lihaoyi** [comment]
Just wanted to add my voice here, the flagship Microsoft Surface Laptop 7 windows laptops are all Windows ARM64 these days. Windows-ARM isn't only an obscure tech for early adopters anymore

**Comment by sagis** [comment]
Adding another request to support ARM builds - many of our customers are moving to ARM based Windows endpoints and not being able to support them like we do on X64 is a major problem

**Comment by thejudge156** [comment]
Adding yet another request.

**Comment by borkdude** [comment]
+1. Even Github Actions supports Windows on Arm now.

**Comment by judepayne** [comment]
+1 for Windows on ARM so that I can use GRAAL to 'complete the set' for my projects

**Comment by haydin505** [comment]
We need this, I cannot use Graal on my Windows which is running on Arm chip Mac

**Comment by borkdude** [comment]
I've heard that Windows on Arm uses an emulator, does it happen to work as a workaround? 
https://learn.microsoft.com/en-us/windows/arm/apps-on-arm-x86-emulation

**Comment by vishvanatarajan** [comment]
+1 GraalVM is one of the major open source software that does not support Windows on ARM.

**Comment by boekabart** [comment]
> I've heard that Windows on Arm uses an emulator, does it happen to work as a workaround? https://learn.microsoft.com/en-us/windows/arm/apps-on-arm-x86-emulation

Performance is kind of a major reason for using Graal. So... I'm going to say, nay.

**Comment by joffrey-bion** [comment]
Have the plans changed at all since then? Maybe this issue should be reopened and reconsidered.

We were considering distributing our new Amper build tool as a GraalVM native image, but if Windows Arm is not supported, this will probably have to wait. It would be a pain to maintain a JVM distribution just for this platform.

**Comment by hohwille** [comment]
> Support for Windows ARM64 is not planned for GraalVM currently.

There is ARM64 support for MacOS (together with x64 support).
Also the same for Linux: https://www.graalvm.org/21.3/reference-manual/native-image/ARM64/index.html
Why not having the same for Windows?
I do not see any reason for rejecting this request.
I fully understand that it may be a matter of priority and take some time but "is not planned" is IMHO the wrong choice here.

**Comment by lewurm** [comment]
> There is ARM64 support for MacOS (together with x64 support).

FWIW, we dropped macOS x86_64 support recently.



Supporting Windwos ARM64 is still not on our agenda, but eventually GraalVM should support this platform, so I'm reopen the issue.  Maybe someone from the community has interest to add support for this platform.

-----

Known workarounds:
- Use Windows x86_64 binaries and go through the Windows' emulator.  It has a certain performance penalty.
- Use WSL and Linux ARM64 binaries.  Near native performance, but depending on the scenario it could be a degraded user experience.

**Comment by fniephaus** [comment]
Thanks for raising this, and apologies for closing the ticket so quickly.

As @lewurm noted, Windows ARM64 support isn’t on our roadmap at the moment. We agree it would be valuable for GraalVM, and we’re open to community contributions.

The existing ARM64 support for Linux and macOS should help inform this effort, but there will be Windows-specific work to investigate. If someone in the community is interested in moving this forward, please reach out. We’re happy to provide guidance and answer questions.

**Comment by lewurm** [comment]
A non-exhaustive list to get a somewhat useful GraalVM support:
* Compiler: An OpenJDK port exists, so as a first step make sure that the Graal compiler (still) works on HotSpot.  In my previous job I had the Graal compiler working on Windows Arm64 with very minimal changes, see https://github.com/openjdk/jdk/pull/972/files most changes are `jaotc`, which can be ignored in the context of this PR
* Native Image: Since [Windows-AArch64 has a slightly different calling convention](https://learn.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=msvc-170) than Linux and macOS, this needs to be reflected in the SubstrateVM backend: https://github.com/oracle/graal/blob/6ebd3c7e2f8518109cbc9ebc21e4198beb8a8ef5/substratevm/src/com.oracle.svm.core.graal.aarch64/src/com/oracle/svm/core/graal/aarch64/SubstrateAArch64RegisterConfig.java#L117  likely many other places that need some fixups in that regard
* Native Image: The objectfile support on Windows must be extended to support AArch64 https://github.com/oracle/graal/tree/master/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff
* Native Image: The way how the toolchain (msvc) is used might needs some updates: https://github.com/oracle/graal/blob/6ebd3c7e2f8518109cbc9ebc21e4198beb8a8ef5/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/CCLinkerInvocation.java#L499-L601
* Native Image: There is a minimal version of Windows API assumed on x86_64.  I'm unsure if the same version is available on Windows Arm.  If not, certain usages in `com.oracle.svm.core.windows` might need some updates https://github.com/oracle/graal/tree/master/substratevm/src/com.oracle.svm.core.windows/src/com/oracle/svm/core/windows
* CI: I don't think github action supports Windows Arm, but in OpenJDK there is at least a cross-build (msvc supports cross compilation): https://github.com/openjdk/jdk/blob/c5159fc9fa0fd81dec629cd821b3411b4a6df967/.github/workflows/main.yml#L363-L376
* `mx`: probably needs some minor platform support.
* Truffle languages: Every language except JavaScript will need a certain amount of language support.

**Comment by linghengqian** [comment]
> CI: I don't think github action supports Windows Arm, 

- As a supplement, there is already a GitHub Actions runner for Windows ARM, which can be invoked by the YAML tag `windows-11-arm`. However, it is Windows 11 rather than Windows Server 2025. Refer to https://github.com/actions/partner-runner-images . 

- But if developers do not have a real Windows ARM device, debugging the code is indeed an extremely frustrating task. Personally, I have never seen anyone in real life purchase Windows ARM devices.

**Comment by hohwille** [comment]
> Known workarounds:
> Use Windows x86_64 binaries and go through the Windows' emulator. It has a certain performance penalty.

So when I run windows x86_64 graalvm on windows@arm64 machine I will be able to generate my release for windows@arm64 instead of x86_64?
So emulation only happens for the build when running graalvm but my end-users having windows@arm64 will not have to use emulation to run my OSS product? That would already be a great workaround for me...

Then I only have the problem that `setup-graalvm` action does not support this:
```
Run graalvm/setup-graalvm@v1
Error: Failed to download graalvm-community-jdk-25.0.1_windows-aarch64_bin.zip. Are you sure java-version: '25.0.1' is correct?
```
https://github.com/devonfw/IDEasy/actions/runs/20812437104/job/59779708855

> Personally, I have never seen anyone in real life purchase Windows ARM devices.

I have requests from such users for my product and therefore I bought such a device myself (Lenovo ThinkPad T14s G6 Qualcomm Snapdragon X Elite) to support this. It is great and battery lasts so much longer than with x86. I love it - but still I cannot provide my OSS product releases for it (https://github.com/devonfw/IDEasy/issues/599).

**Comment by lewurm** [comment]
> So when I run windows x86_64 graalvm on windows@arm64 machine I will be able to generate my release for windows@arm64 instead of x86_64?

No, you would run a Windows x86_64 GraalVM on Windows Arm64 machine, but also generate a binary that is targeting x86_64.  The resulting binary should also work with the emulator.  If not, please open another issue.


Cross compilation in general is not supported by Native Image https://github.com/oracle/graal/issues/407
